### PR TITLE
Enable no implicit coercion eslint rule and pre-apply fixes

### PR DESCRIPTION
**Pull request by tylerbutler**
From the [rule documentation](https://eslint.org/docs/rules/no-implicit-coercion):

> In JavaScript, there are a lot of different ways to convert value types. Some of them might be hard to read and understand.

### CODE REVIEW

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 1]
```diff
@@ -305,14 +305,14 @@ export class DataStores implements IDisposable {
...
-        assert(!!localContext, 0x15f /* "Could not find unbound context to bind" */);
+        assert(Boolean(localContext), 0x15f /* "Could not find unbound context to bind" */);
 
         /**
          * If the container is not detached, it is globally visible to all clients. This data store should also be
          * globally visible. Move it to attaching state and send an "attach" op for it.
          * If the container is detached, this data store will be part of the summary that makes the container attached.
          */
-        if (this.runtime.attachState !== AttachState.Detached) {
+        if (localContext && this.runtime.attachState !== AttachState.Detached) {
```
Comment (ChumpChief) [inline packages/runtime/container-runtime/src/dataStores.ts:315]: Above, prefer to assert localContext !== undefined, then you don't need this extra check

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 2]
```diff
@@ -376,13 +376,19 @@ export class DataStores implements IDisposable {
     public resubmitDataStoreOp(content: any, localOpMetadata: unknown) {
         const envelope = content as IEnvelope;
         const context = this.contexts.get(envelope.address);
+        // The implicit coercion below also informs TypeScript that context cannot be null. Removing the implicit
+        // coercion here would make the subsequent code more complex.
+        // eslint-disable-next-line no-implicit-coercion
         assert(!!context, 0x160 /* "There should be a store context for the op" */);
```
Comment (ChumpChief) [inline packages/runtime/container-runtime/src/dataStores.ts:382]: Most of these `!!thing` usages should really be `thing !== undefined`, which is a stronger check anyway.

**Review by ChumpChief — COMMENTED** [review]
I might recommend turning on strict-boolean-expressions first if possible, since most of these Boolean coercions would be better solved by making them strict.

File: packages/runtime/container-runtime/src/dataStores.ts (hunk 2 above)
Comment (tylerbutler) [reply packages/runtime/container-runtime/src/dataStores.ts:382]: Ohhhh, that's a way better fix... I'll update where I can. I hate having these comments everywhere.

File: examples/data-objects/shared-text/src/client-ui-lib/controls/flowView.ts [hunk 3]
```diff
@@ -2661,7 +2661,7 @@ export class FlowView extends ui.Component {
                 }
             }
 
-            const lc = !!(tileInfo.tile as Paragraph.IParagraphMarker).listCache;
+            const lc = Boolean((tileInfo.tile as Paragraph.IParagraphMarker).listCache);
```
Comment (tylerbutler) [inline examples/data-objects/shared-text/src/client-ui-lib/controls/flowView.ts:2664]: Could this be:

```suggestion
            const lc = (tileInfo.tile as Paragraph.IParagraphMarker)?.listCache);
```

**Comment by tylerbutler** [comment]
Converted this to a draft since I am going to work on strict-boolean-expressions first.

**Pull request closed without merging**